executor/write.go

```diff
@@ -44,114 +44,118 @@ var (
+			return false, false, 0, errors.Trace(err)
 		} else {
-			if mysql.HasOnUpdateNowFlag(col.Flag) && modified[i] {
-				// It's for "UPDATE t SET ts = ts" and ts is a timestamp.
-				onUpdateSpecified[i] = true
+			if cmp != 0 {
+				changed = true
+				modified[i] = true
+				// Rebase auto increment id if the field is changed.
+				if mysql.HasAutoIncrementFlag(col.Flag) {
+					if err := t.RebaseAutoID(ctx, newData[i].GetInt64(), true); err != nil {
+						return false, false, 0, errors.Trace(err)
+					}
+				}
+				if col.IsPKHandleColumn(t.Meta()) {
+					handleChanged = true
+					newHandle = newData[i].GetInt64()
+				}
+			} else {
+				if mysql.HasOnUpdateNowFlag(col.Flag) && modified[i] {
+					// It's for "UPDATE t SET ts = ts" and ts is a timestamp.
+					onUpdateSpecified[i] = true
+				}
+				modified[i] = false
 			}
-			modified[i] = false
 		}
 	}
 
-	// Check the not-null constraints.
-	err := table.CheckNotNull(t.Cols(), newData)
-	if err != nil {
-		return false, handleChanged, newHandle, errors.Trace(err)
-	}
-
+	// If no changes, nothing to do, return directly.
 	if !changed {
 		// See https://dev.mysql.com/doc/refman/5.7/en/mysql-real-connect.html  CLIENT_FOUND_ROWS
 		if ctx.GetSessionVars().ClientCapability&mysql.ClientFoundRows > 0 {
 			sc.AddAffectedRows(1)
 		}
-		return false, handleChanged, newHandle, nil
+		return false, false, 0, nil
 	}
 
-	// Fill values into on-update-now fields, only if they are really changed.
+	// 4. Fill values into on-update-now fields, only if they are really changed.
 	for i, col := range t.Cols() {
 		if mysql.HasOnUpdateNowFlag(col.Flag) && !modified[i] && !onUpdateSpecified[i] {
-			v, errGT := expression.GetTimeValue(ctx, strings.ToUpper(ast.CurrentTimestamp), col.Tp, col.Decimal)
-			if errGT != nil {
-				return false, handleChanged, newHandle, errors.Trace(errGT)
+			if v, err := expression.GetTimeValue(ctx, strings.ToUpper(ast.CurrentTimestamp), col.Tp, col.Decimal); err != nil {
+				return false, false, 0, errors.Trace(err)
+			} else {
+				newData[i] = v
+				modified[i] = true
 			}
-			newData[i] = v
-			modified[i] = true
 		}
 	}
 
+	// 5. If handle changed, remove the old then add the new record, otherwise update the record.
+	var err error
 	if handleChanged {
 		skipHandleCheck := false
 		if sc.DupKeyAsWarning {
-			// if the new handle exists. `UPDATE IGNORE` will avoid removing record, and do nothing.
+			// For `UPDATE IGNORE`/`INSERT IGNORE ON DUPLICATE KEY UPDATE`
+			// If the new handle exists, this will avoid to remove the record.
 			err = tables.CheckHandleExists(ctx, t, newHandle, newData)
 			if err != nil {
 				return false, handleChanged, newHandle, errors.Trace(err)
 			}
 			skipHandleCheck = true
 		}
-		err = t.RemoveRecord(ctx, h, oldData)
-		if err != nil {
-			return false, handleChanged, newHandle, errors.Trace(err)
+		if err := t.RemoveRecord(ctx, h, oldData); err != nil {
```

Comment:
This error shadows the outer one?

Reply:
Oh, I'll check it as soon as I fix the `make check` on my laptop.